File: go/2316.go
package main

func countPairs(n int, edges [][]int) (ans int64) {
	g := make([][]int, n)
	for _, edge := range edges {
		x, y := edge[0], edge[1]
		g[x] = append(g[x], y)
		g[y] = append(g[y], x)
	}
	v := make([]bool, n)
	var dfs func(int) int
	dfs = func(i int) (ans int) {
		q := make([]int, 1)
		q[0] = i
		for len(q) > 0 {
			x := q[len(q)-1]
			q = q[:len(q)-1]
			if !v[x] {
				ans++
				v[x] = true
				for _, y := range g[x] {
					q = append(q, y)
				}
			}
		}
		return
	}
	total := 0
	for i := 0; i < n; i++ {
		s := dfs(i)
		ans += int64(total * s)
		total += s
	}
	return ans
}

func main() {

}

File: python/2316.py
from typing import List

class Solution:
    def countPairs(self, n: int, edges: List[List[int]]) -> int:
        g = [[] for _ in range(n)]
        for x, y in edges:
            g[x].append(y)
            g[y].append(x)
        v = [False] * n
        def dfs(n: int) -> int:
            ans = 0
            q = [n]
            while q:
                x = q.pop()
                if not v[x]:
                    ans += 1
                    v[x] = True
                    for y in g[x]:
                        q.append(y)
            return ans
        ans = 0
        total = 0
        for i in range(n):
            s = dfs(i)
            ans += s * total
            total += s
        return ans
            

s = Solution()
n = 7
edges = [[0,2],[0,5],[2,4],[1,6],[5,4]]
print(s.countPairs(n,edges))
        
